packages/web3torrent/src/library/web3torrent-lib.ts

```diff
@@ -367,85 +365,90 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
       switch (command) {
         case PaidStreamingExtensionNotices.STOP: // synonymous with a prompt for a payment
           if (torrent.paused) {
-            // We currently treat pausing torrent as canceling downloads
             await this.closeDownloadingChannels(torrent);
           } else if (!torrent.done) {
-            const channelId = data;
-            await this.paymentChannelClient.makePayment(
-              channelId,
-              WEI_PER_BYTE.mul(BUFFER_REFILL_RATE).toString()
-            );
-
-            let balance: string;
-            const channel = this.paymentChannelClient.channelCache[channelId];
-
-            if (channel) {
-              balance = channel.beneficiaryBalance;
-            } else {
-              balance = 'unknown';
-            }
-
-            log(
-              `attempted to make payment for channel ${channelId}, beneficiaryBalance: ${balance}`
-            );
+            await this.makePayment(torrent, wire);
           }
           break;
         case PaidStreamingExtensionNotices.START:
           this.jumpStart(torrent, wire);
           break;
         case PaidStreamingExtensionNotices.MESSAGE:
-          message = JSON.parse(data.message);
-          if (message.recipient === this.pseAccount) {
-            await this.paymentChannelClient.pushMessage(message);
-          }
-          break;
-        default:
+          await this.paymentChannelClient.pushMessage(data);
           break;
       }
       this.emit(ClientEvents.TORRENT_NOTICE, {torrent, wire, command, data});
     });
 
     torrent.on(TorrentEvents.DONE, async () => {
-      log('Torrent DONE!');
+      log('<< Torrent DONE!', torrent, this.peersList[torrent.infoHash]);
       this.emit(ClientEvents.TORRENT_DONE, {torrent});
       await this.closeDownloadingChannels(torrent);
     });
 
     torrent.on(TorrentEvents.ERROR, err => {
-      log('ERROR: > ', err);
+      log('Torrent ERROR: ', err, torrent);
       this.emit(ClientEvents.TORRENT_ERROR, {torrent, err});
     });
     torrent.usingPaidStreaming = true;
 
     return torrent;
   }
 
+  /**
+   * Define the amount to pay, and makes the payment
+   */
+  protected async makePayment(torrent: PaidStreamingTorrent, wire: PaidStreamingWire) {
+    const {peerChannelId, peerAccount} = wire.paidStreamingExtension;
+    log(`<< STOP ${peerAccount} - About to pay`, torrent);
+    let amountToPay = WEI_PER_BYTE.mul(BUFFER_REFILL_RATE).toString();
+    // On each wire, the algorithm tries to download the uneven piece (which is always the last piece)
+
+    const lastPieceReservations: PaidStreamingWire[] =
+      torrent._reservations[torrent.pieces.length - 1];
+    const lastPieceReservationsIncludesThisWire =
+      lastPieceReservations &&
+      lastPieceReservations.find(
+        wire => wire && wire.paidStreamingExtension.peerAccount === peerAccount
+      );
+    if (torrent.downloaded === 0 && lastPieceReservationsIncludesThisWire) {
```

Comment:
It would be cleaner if we can wrap the Line 410-417 logic in a separate function, e.g., `isLastPiece(torrent, wire)`

Reply:
:+1: